File: Backends/LWJGL3Backend/src/main/java/org/tobi29/scapes/engine/backends/lwjgl3/opengl/VBO.java
package org.tobi29.scapes.engine.backends.lwjgl3.opengl;

import java8.util.Optional;
import org.lwjgl.opengl.*;
import org.tobi29.scapes.engine.ScapesEngine;
import org.tobi29.scapes.engine.graphics.GL;
import org.tobi29.scapes.engine.graphics.ModelAttribute;
import org.tobi29.scapes.engine.graphics.VertexType;
import org.tobi29.scapes.engine.utils.Streams;
import org.tobi29.scapes.engine.utils.math.FastMath;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.List;

final class VBO {
    private final ScapesEngine engine;
    private final int stride;
    private final List<ModelAttributeData> attributes = new ArrayList<>();
    private Optional<ByteBuffer> data;
    private int vertexID;
    private boolean stored;

    public VBO(ScapesEngine engine, List<ModelAttribute> attributes,
            int length) {
        this.engine = engine;
        int stride = 0;
        for (ModelAttribute attribute : attributes) {
            if (attribute.length() != length * attribute.size()) {
                throw new IllegalArgumentException(
                        "Inconsistent attribute data length");
            }
            this.attributes.add(new ModelAttributeData(attribute, stride));
            attribute.setOffset(stride);
            int size = attribute.size() * attribute.vertexType().bytes();
            stride += (size - 1 | 0x03) + 1;
        }
        this.stride = stride;
        ByteBuffer vertexBuffer = engine.allocate(length * stride);
        Streams.forEach(attributes,
                attribute -> addToBuffer(attribute, length, vertexBuffer));
        data = Optional.of(vertexBuffer);
    }

    public ScapesEngine engine() {
        return engine;
    }

    public int stride() {
        return stride;
    }

    public void replaceBuffer(GL gl, ByteBuffer buffer) {
        assert stored;
        gl.check();
        GL15.glBindBuffer(GL15.GL_ARRAY_BUFFER, vertexID);
        GL15.glBufferData(GL15.GL_ARRAY_BUFFER, buffer.capacity(),
                GL15.GL_STREAM_DRAW);
        GL15.glBufferSubData(GL15.GL_ARRAY_BUFFER, 0, buffer);
    }

    private void storeAttribute(GL gl, ModelAttributeData attribute) {
        gl.check();
        GL20.glEnableVertexAttribArray(attribute.id);
        switch (attribute.vertexType) {
            case FLOAT:
                GL20.glVertexAttribPointer(attribute.id, attribute.size,
                        GL11.GL_FLOAT, attribute.normalized, stride,
                        attribute.offset);
                break;
            case HALF_FLOAT:
                GL20.glVertexAttribPointer(attribute.id, attribute.size,
                        GL30.GL_HALF_FLOAT, attribute.normalized, stride,
                        attribute.offset);
                break;
            case BYTE:
                GL20.glVertexAttribPointer(attribute.id, attribute.size,
                        GL11.GL_BYTE, attribute.normalized, stride,
                        attribute.offset);
                break;
            case UNSIGNED_BYTE:
                GL20.glVertexAttribPointer(attribute.id, attribute.size,
                        GL11.GL_UNSIGNED_BYTE, attribute.normalized, stride,
                        attribute.offset);
                break;
            case SHORT:
                GL20.glVertexAttribPointer(attribute.id, attribute.size,
                        GL11.GL_SHORT, attribute.normalized, stride,
                        attribute.offset);
                break;
            case UNSIGNED_SHORT:
                GL20.glVertexAttribPointer(attribute.id, attribute.size,
                        GL11.GL_UNSIGNED_SHORT, attribute.normalized, stride,
                        attribute.offset);
                break;
            default:
                throw new IllegalArgumentException("Unknown vertex type!");
        }
        GL33.glVertexAttribDivisor(attribute.id, attribute.divisor);
    }

    private void addToBuffer(ModelAttribute attribute, int vertices,
            ByteBuffer buffer) {
        if (attribute.floatArray() == null) {
            switch (attribute.vertexType()) {
                case BYTE:
                case UNSIGNED_BYTE:
                    for (int i = 0; i < vertices; i++) {
                        int is = i * attribute.size();
                        buffer.position(attribute.offset() + i * stride);
                        for (int j = 0; j < attribute.size(); j++) {
                            int ij = is + j;
                            buffer.put(attribute.byteArray()[ij]);
                        }
                    }
                    break;
                case SHORT:
                case UNSIGNED_SHORT:
                    for (int i = 0; i < vertices; i++) {
                        int is = i * attribute.size();
                        buffer.position(attribute.offset() + i * stride);
                        for (int j = 0; j < attribute.size(); j++) {
                            int ij = is + j << 1;
                            buffer.putShort((short) (
                                    attribute.byteArray()[ij + 1] << 8 |
                                            attribute.byteArray()[ij]));
                        }
                    }
                    break;
                default:
                    throw new IllegalArgumentException(
                            "Invalid array in vao attribute!");
            }
        } else {
            switch (attribute.vertexType()) {
                case FLOAT:
                    for (int i = 0; i < vertices; i++) {
                        int is = i * attribute.size();
                        buffer.position(attribute.offset() + i * stride);
                        for (int j = 0; j < attribute.size(); j++) {
                            int ij = is + j;
                            buffer.putFloat(attribute.floatArray()[ij]);
                        }
                    }
                    break;
                case HALF_FLOAT:
                    for (int i = 0; i < vertices; i++) {
                        int is = i * attribute.size();
                        buffer.position(attribute.offset() + i * stride);
                        for (int j = 0; j < attribute.size(); j++) {
                            int ij = is + j;
                            buffer.putShort(FastMath.convertFloatToHalf(
                                    attribute.floatArray()[ij]));
                        }
                    }
                    break;
                case BYTE:
                    if (attribute.normalized()) {
                        for (int i = 0; i < vertices; i++) {
                            int is = i * attribute.size();
                            buffer.position(attribute.offset() + i * stride);
                            for (int j = 0; j < attribute.size(); j++) {
                                int ij = is + j;
                                buffer.put((byte) FastMath
                                        .round(attribute.floatArray()[ij] *
                                                127.0f));
                            }
                        }
                    } else {
                        for (int i = 0; i < vertices; i++) {
                            int is = i * attribute.size();
                            buffer.position(attribute.offset() + i * stride);
                            for (int j = 0; j < attribute.size(); j++) {
                                int ij = is + j;
                                buffer.put((byte) FastMath
                                        .round(attribute.floatArray()[ij]));
                            }
                        }
                    }
                    break;
                case UNSIGNED_BYTE:
                    if (attribute.normalized()) {
                        for (int i = 0; i < vertices; i++) {
                            int is = i * attribute.size();
                            buffer.position(attribute.offset() + i * stride);
                            for (int j = 0; j < attribute.size(); j++) {
                                int ij = is + j;
                                buffer.put((byte) FastMath
                                        .round(attribute.floatArray()[ij] *
                                                255.0f));
                            }
                        }
                    } else {
                        for (int i = 0; i < vertices; i++) {
                            int is = i * attribute.size();
                            buffer.position(attribute.offset() + i * stride);
                            for (int j = 0; j < attribute.size(); j++) {
                                int ij = is + j;
                                buffer.put((byte) FastMath
                                        .round(attribute.floatArray()[ij]));
                            }
                        }
                    }
                    break;
                case SHORT:
                    if (attribute.normalized()) {
                        for (int i = 0; i < vertices; i++) {
                            int is = i * attribute.size();
                            buffer.position(attribute.offset() + i * stride);
                            for (int j = 0; j < attribute.size(); j++) {
                                int ij = is + j;
                                buffer.putShort((short) FastMath
                                        .round(attribute.floatArray()[ij] *
                                                32768.0f));
                            }
                        }
                    } else {
                        for (int i = 0; i < vertices; i++) {
                            int is = i * attribute.size();
                            buffer.position(attribute.offset() + i * stride);
                            for (int j = 0; j < attribute.size(); j++) {
                                int ij = is + j;
                                buffer.putShort((short) FastMath
                                        .round(attribute.floatArray()[ij]));
                            }
                        }
                    }
                    break;
                case UNSIGNED_SHORT:
                    if (attribute.normalized()) {
                        for (int i = 0; i < vertices; i++) {
                            int is = i * attribute.size();
                            buffer.position(attribute.offset() + i * stride);
                            for (int j = 0; j < attribute.size(); j++) {
                                int ij = is + j;
                                buffer.putShort((short) FastMath
                                        .round(attribute.floatArray()[ij] *
                                                65535.0f));
                            }
                        }
                    } else {
                        for (int i = 0; i < vertices; i++) {
                            int is = i * attribute.size();
                            buffer.position(attribute.offset() + i * stride);
                            for (int j = 0; j < attribute.size(); j++) {
                                int ij = is + j;
                                buffer.putShort((short) FastMath
                                        .round(attribute.floatArray()[ij]));
                            }
                        }
                    }
                    break;
                default:
                    throw new IllegalArgumentException(
                            "Invalid array in vao attribute!");
            }
        }
    }

    public boolean canStore() {
        return data.isPresent();
    }

    public void store(GL gl, boolean weak) {
        assert !stored;
        stored = true;
        gl.check();
        ByteBuffer data = this.data.get();
        data.rewind();
        vertexID = GL15.glGenBuffers();
        GL15.glBindBuffer(GL15.GL_ARRAY_BUFFER, vertexID);
        GL15.glBufferData(GL15.GL_ARRAY_BUFFER, data, GL15.GL_STATIC_DRAW);
        Streams.forEach(attributes, attribute -> storeAttribute(gl, attribute));
        if (weak) {
            this.data = Optional.empty();
        }
    }

    public void dispose(GL gl) {
        assert stored;
        stored = false;
        gl.check();
        GL15.glDeleteBuffers(vertexID);
    }

    public void reset() {
        stored = false;
    }

    private static final class ModelAttributeData {
        private final VertexType vertexType;
        private final int id, size, offset, divisor;
        private final boolean normalized;

        private ModelAttributeData(ModelAttribute attribute, int offset) {
            this.offset = offset;
            vertexType = attribute.vertexType();
            id = attribute.id();
            size = attribute.size();
            normalized = attribute.normalized();
            divisor = attribute.divisor();
        }
    }
}

File: LWJGL3BackendJVM/src/main/kotlin/org/tobi29/scapes/engine/backends/lwjgl3/opengl/VBO.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.engine.backends.lwjgl3.opengl

import org.tobi29.scapes.engine.graphics.GL
import org.tobi29.scapes.engine.graphics.GraphicsObjectSupplier
import org.tobi29.scapes.engine.graphics.ModelAttribute
import org.tobi29.scapes.engine.graphics.VertexType
import org.tobi29.math.FastMath
import org.tobi29.stdex.assert
import org.tobi29.io.ByteViewE
import org.tobi29.io.ByteViewRO
import kotlin.math.round

internal class VBO(val gos: GraphicsObjectSupplier,
                   attributes: List<ModelAttribute>,
                   length: Int) {
    private val stride: Int
    private val attributes = ArrayList<ModelAttributeData>()
    private var data: ByteViewRO? = null
    private var vertexID = 0
    private var stored = false

    init {
        var stride = 0
        for (attribute in attributes) {
            if (attribute.length != length * attribute.size) {
                throw IllegalArgumentException(
                        "Inconsistent attribute data length")
            }
            this.attributes.add(ModelAttributeData(attribute, stride))
            attribute.offset = stride
            val size = attribute.size * attribute.vertexType.bytes
            stride += (size - 1 or 0x03) + 1
        }
        this.stride = stride
        val vertexBuffer = gos.container.allocateNative(length * stride)
        attributes.forEach { addToBuffer(it, length, vertexBuffer) }
        data = vertexBuffer
    }

    fun stride(): Int {
        return stride
    }

    fun replaceBuffer(gl: GL,
                      buffer: ByteViewRO) {
        gl.check()
        data = buffer
        if (!stored) return
        glBindBuffer(GL_ARRAY_BUFFER, vertexID)
        glBufferData(GL_ARRAY_BUFFER, buffer.size, GL_STREAM_DRAW)
        glBufferSubData(GL_ARRAY_BUFFER, 0, buffer)
    }

    private fun storeAttribute(gl: GL,
                               attribute: ModelAttributeData) {
        gl.check()
        glEnableVertexAttribArray(attribute.id)
        if (attribute.integer) {
            when (attribute.vertexType) {
                VertexType.FLOAT -> glVertexAttribIPointer(attribute.id,
                        attribute.size,
                        GL_FLOAT, stride, attribute.offset)
                VertexType.HALF_FLOAT -> glVertexAttribIPointer(
                        attribute.id, attribute.size,
                        GL_HALF_FLOAT, stride, attribute.offset)
                VertexType.BYTE -> glVertexAttribIPointer(attribute.id,
                        attribute.size,
                        GL_BYTE, stride, attribute.offset)
                VertexType.UNSIGNED_BYTE -> glVertexAttribIPointer(
                        attribute.id, attribute.size,
                        GL_UNSIGNED_BYTE, stride,
                        attribute.offset)
                VertexType.SHORT -> glVertexAttribIPointer(attribute.id,
                        attribute.size,
                        GL_SHORT, stride, attribute.offset)
                VertexType.UNSIGNED_SHORT -> glVertexAttribIPointer(
                        attribute.id, attribute.size,
                        GL_UNSIGNED_SHORT, stride,
                        attribute.offset)
                else -> throw IllegalArgumentException("Unknown vertex type!")
            }
        } else {
            when (attribute.vertexType) {
                VertexType.FLOAT -> glVertexAttribPointer(attribute.id,
                        attribute.size,
                        GL_FLOAT, attribute.normalized, stride,
                        attribute.offset)
                VertexType.HALF_FLOAT -> glVertexAttribPointer(
                        attribute.id, attribute.size,
                        GL_HALF_FLOAT, attribute.normalized, stride,
                        attribute.offset)
                VertexType.BYTE -> glVertexAttribPointer(attribute.id,
                        attribute.size,
                        GL_BYTE, attribute.normalized, stride,
                        attribute.offset)
                VertexType.UNSIGNED_BYTE -> glVertexAttribPointer(
                        attribute.id, attribute.size,
                        GL_UNSIGNED_BYTE, attribute.normalized, stride,
                        attribute.offset)
                VertexType.SHORT -> glVertexAttribPointer(attribute.id,
                        attribute.size,
                        GL_SHORT, attribute.normalized, stride,
                        attribute.offset)
                VertexType.UNSIGNED_SHORT -> glVertexAttribPointer(
                        attribute.id, attribute.size,
                        GL_UNSIGNED_SHORT, attribute.normalized,
                        stride, attribute.offset)
                else -> throw IllegalArgumentException("Unknown vertex type!")
            }
        }
        glVertexAttribDivisor(attribute.id, attribute.divisor)
    }

    private fun addToBuffer(attribute: ModelAttribute,
                            vertices: Int,
                            buffer: ByteViewE) {
        val floatArray = attribute.floatArray
        if (floatArray == null) {
            val byteArray = attribute.byteArray ?: throw IllegalArgumentException(
                    "Attribute contains no data")
            when (attribute.vertexType) {
                VertexType.BYTE, VertexType.UNSIGNED_BYTE ->
                    buffer.storeBytes({ byteArray[it].toByte() },
                            vertices, attribute.offset, attribute.size, stride)
                VertexType.SHORT, VertexType.UNSIGNED_SHORT ->
                    buffer.storeShorts({ byteArray[it].toShort() },
                            vertices, attribute.offset, attribute.size, stride)
                else -> throw IllegalArgumentException(
                        "Invalid array in vao attribute!")
            }
        } else {
            when (attribute.vertexType) {
                VertexType.FLOAT ->
                    buffer.storeFloats({ floatArray[it] },
                            vertices, attribute.offset, attribute.size, stride)
                VertexType.HALF_FLOAT ->
                    buffer.storeShorts(
                            { FastMath.convertFloatToHalf(floatArray[it]) },
                            vertices, attribute.offset, attribute.size, stride)
                VertexType.BYTE -> if (attribute.normalized) {
                    buffer.storeBytes(
                            { round(floatArray[it] * 127.0f).toByte() },
                            vertices, attribute.offset, attribute.size, stride)
                } else {
                    buffer.storeBytes(
                            { round(floatArray[it]).toByte() },
                            vertices, attribute.offset, attribute.size, stride)
                }
                VertexType.UNSIGNED_BYTE -> if (attribute.normalized) {
                    buffer.storeBytes(
                            { round(floatArray[it] * 255.0f).toByte() },
                            vertices, attribute.offset, attribute.size, stride)
                } else {
                    buffer.storeBytes(
                            { round(floatArray[it]).toByte() },
                            vertices, attribute.offset, attribute.size, stride)
                }
                VertexType.SHORT -> if (attribute.normalized) {
                    buffer.storeShorts(
                            { round(floatArray[it] * 32768.0f).toShort() },
                            vertices, attribute.offset, attribute.size, stride)
                } else {
                    buffer.storeShorts(
                            { round(floatArray[it]).toShort() },
                            vertices, attribute.offset, attribute.size, stride)
                }
                VertexType.UNSIGNED_SHORT -> if (attribute.normalized) {
                    buffer.storeShorts(
                            { round(floatArray[it] * 65535.0f).toShort() },
                            vertices, attribute.offset, attribute.size, stride)
                } else {
                    buffer.storeShorts(
                            { round(floatArray[it]).toShort() },
                            vertices, attribute.offset, attribute.size, stride)
                }
                else -> throw IllegalArgumentException(
                        "Invalid array in vao attribute!")
            }
        }
    }

    fun canStore(): Boolean {
        return data != null
    }

    fun store(gl: GL,
              weak: Boolean) {
        assert { !stored }
        val data = data ?: throw IllegalStateException(
                "VBO cannot be stored anymore")
        stored = true
        gl.check()
        vertexID = glGenBuffers()
        glBindBuffer(GL_ARRAY_BUFFER, vertexID)
        glBufferData(GL_ARRAY_BUFFER, data, GL_STATIC_DRAW)
        attributes.forEach { storeAttribute(gl, it) }
        if (weak) {
            this.data = null
        }
    }

    fun dispose(gl: GL) {
        assert { stored }
        stored = false
        gl.check()
        glDeleteBuffers(vertexID)
    }

    fun reset() {
        stored = false
    }

    private class ModelAttributeData(attribute: ModelAttribute,
                                     val offset: Int) {
        val vertexType = attribute.vertexType
        val id = attribute.id
        val size = attribute.size
        val divisor = attribute.divisor
        val normalized = attribute.normalized
        val integer = attribute.byteArray != null
    }
}

private inline fun ByteViewE.storeBytes(source: (Int) -> Byte,
                                        vertices: Int,
                                        offset: Int,
                                        size: Int,
                                        stride: Int) =
        store({ i, j -> setByte(j, source(i)) }, vertices, offset, size,
                stride, 1)

private inline fun ByteViewE.storeShorts(source: (Int) -> Short,
                                         vertices: Int,
                                         offset: Int,
                                         size: Int,
                                         stride: Int) =
        store({ i, j -> setShort(j, source(i)) }, vertices, offset, size,
                stride, 2)

private inline fun ByteViewE.storeInts(source: (Int) -> Int,
                                       vertices: Int,
                                       offset: Int,
                                       size: Int,
                                       stride: Int) =
        store({ i, j -> setInt(j, source(i)) }, vertices, offset, size,
                stride, 4)

private inline fun ByteViewE.storeFloats(source: (Int) -> Float,
                                         vertices: Int,
                                         offset: Int,
                                         size: Int,
                                         stride: Int) =
        store({ i, j -> setFloat(j, source(i)) }, vertices, offset, size,
                stride, 4)

private inline fun store(copy: (Int, Int) -> Unit,
                         vertices: Int,
                         offset: Int,
                         size: Int,
                         stride: Int,
                         typeStride: Int) {
    for (i in 0 until vertices) {
        val o = i * size
        var k = offset + i * stride
        for (j in o until o + size) {
            copy(j, k)
            k += typeStride
        }
    }
}
